include/oneapi/dpl/pstl/hetero/histogram_impl_hetero.h

```diff
@@ -0,0 +1,174 @@
+    template <typename _Handler>
+    auto
+    prepare_device_binhash(_Handler& __cgh) const
+    {
+        return __bin_hash;
+    }
+};
+
+// Augmentation for binhash function which stores a range of dynamic memory to determine bin mapping
+template <typename _Range, typename _BufferHolder>
+struct __binhash_manager_custom_boundary
+    : public __binhash_manager_base<oneapi::dpl::__par_backend_hetero::__custom_boundary_range_binhash<_Range>>
+{
+    using _bin_hash_type = oneapi::dpl::__par_backend_hetero::__custom_boundary_range_binhash<_Range>;
+    using _base_type =
+        __binhash_manager_base<oneapi::dpl::__par_backend_hetero::__custom_boundary_range_binhash<_Range>>;
+    using _extra_memory_type = __value_t<_Range>;
+    // __buffer_holder is required to keep the sycl buffer alive until the kernel has been completed (waited on)
+    _BufferHolder __buffer_holder;
+
+    __binhash_manager_custom_boundary(_bin_hash_type&& __bin_hash_, _BufferHolder&& __buffer_holder_)
+        : _base_type(::std::move(__bin_hash_)), __buffer_holder(::std::move(__buffer_holder_))
+    {
+    }
+
+    auto
+    get_required_SLM_elements() const
+    {
+        return this->__bin_hash.__boundaries.size();
+    }
+
+    template <typename _Handler>
+    auto
+    prepare_device_binhash(_Handler& __cgh) const
+    {
+        oneapi::dpl::__ranges::__require_access(__cgh, this->__bin_hash.__boundaries);
+        return _base_type::prepare_device_binhash(__cgh);
+    }
+};
+
+template <typename _BinHash>
+auto
+__make_binhash_manager(_BinHash&& __bin_hash)
+{
+    return __binhash_manager_base(::std::forward<_BinHash>(__bin_hash));
+}
+
+template <typename _RandomAccessIterator>
+auto
+__make_binhash_manager(oneapi::dpl::__internal::__custom_boundary_binhash<_RandomAccessIterator>&& __bin_hash)
+{
+    auto __buffer_lifetime_holder =
+        oneapi::dpl::__ranges::__get_sycl_range<oneapi::dpl::__par_backend_hetero::access_mode::read,
+                                                _RandomAccessIterator>();
+    auto __range_holder = __buffer_lifetime_holder(__bin_hash.__boundary_first, __bin_hash.__boundary_last);
+    auto __bin_hash_range =
+        oneapi::dpl::__par_backend_hetero::__custom_boundary_range_binhash{__range_holder.all_view()};
+    return __binhash_manager_custom_boundary{::std::move(__bin_hash_range), ::std::move(__buffer_lifetime_holder)};
+}
+
+template <typename _Name>
+struct __hist_fill_zeros_wrapper
+{
+};
+
+template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _Size, typename _BinHash,
+          typename _RandomAccessIterator2>
+oneapi::dpl::__internal::__enable_if_hetero_execution_policy<_ExecutionPolicy>
+__pattern_histogram(_ExecutionPolicy&& __exec, _RandomAccessIterator1 __first, _RandomAccessIterator1 __last,
+                    _Size __num_bins, _BinHash&& __func, _RandomAccessIterator2 __histogram_first)
+{
+    //If there are no histogram bins there is nothing to do
+    if (__num_bins > 0)
```

Comment:
Just minor comment for the future.
There is recommendation to keep the sorter (or no op) code branch as first branch of "if else" operator. It improves readability (easier to control code branching, no indentation for many lines), and sometimes makes smaller the file size.